File: Boggle.cpp
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <algorithm>
#include <cstring>
#include <cctype>
using namespace std;


#define LEN 18

struct WNode
{
	char word[LEN];
	WNode* next;
};

class Boogle
{
public:
	Boogle(void){};
//	Boogle(Boogle &other);
	int Randomize(const char* dice[]);
	int Customize(void);
	int Print(int score);
	WNode* FindWords(WNode* dict);
	WNode* FakeWords(WNode* dict);
	int PrintWords(void);
	bool isValidWord(char* word);

	int maxP;

private:
	char letters[17];
};




int validate(const char* word, WNode* wordList, WNode* prev);
WNode* addWord(const char* word, WNode* list);

bool checkGuess(const char* word, WNode* sList, WNode* pList)
{
	auto WNode* traverser = pList;

	while(traverser)
		{
		if(0== strcmp(word, traverser->word)) return false;
		traverser = traverser->next;
		}

	traverser = sList;
	while(traverser)
		{
		if(0 == strcmp(word, traverser->word)) return true;
		traverser = traverser->next;
		}
	return false;
}


WNode* LoadDict(const char* fname)
{
	auto ifstream infile;
	auto int i = 0;
	auto char word[LEN];
	auto WNode* headPtr = new WNode;
	auto WNode* adder = headPtr;

	infile.open(fname);
	infile >> adder->word;
	adder->next = new WNode;

	while(infile >> adder->next->word)
		{
		adder = adder->next;
		adder->next = new WNode;
		}
	delete adder->next;
	adder->next = NULL;
	infile.close();
if (headPtr == NULL) cout << "Dictionary failed to load\n";
	return headPtr;
}

void DestroyList(WNode* head)
{
	auto WNode* trail = head;
	while(head)
		{
		head = head->next;
		delete trail;
		trail = head;
		}
}



void tester(WNode* d)
{
	int i;
	for(i = 0; i < 20; i++)
		{
		cout << d->word << endl;
		d = d->next;
		}

	cout << "\nLast Word:\n  ";
	while(d->next != NULL) d = d->next;

	cout << d->word << endl;
}


int MaxScore(WNode* wList)
{
	auto WNode* traverser = wList;
	auto int total = 0;

	while(traverser)
		{
		total += (strlen(traverser->word) - 2);
		traverser = traverser->next;
		}

	return total;
}



int main(void)
{
	auto const char* dice[16] = {"VITEGN"
								,"ACESLR"
								,"VAZEDN"
								,"ICATAO"
								,"NODUKT"
								,"ENIPHS"
								,"ORIFBX"
								,"KULEGY"
								,"EYIEHF"
								,"ESUTLP"
								,"EWOSDN"
								,"PECADM"
								,"ALIBTY"
								,"SAHOMR"
								,"JABOMQ"
								,"URIGLW"};


	srand((unsigned)time(NULL));
	auto WNode* wordList = LoadDict("dict.txt");
//	tester(wordList);//yep it looks like it worked

	//create grid
	auto Boogle grid;

	//randomize, solve, & print
	grid.Randomize(dice);
	auto WNode* solvedList = grid.FindWords(wordList);
//auto WNode* solvedList = grid.FakeWords(wordList);

	auto WNode* pList = NULL;
	auto WNode* tbag;
	auto char datChar;
	auto char userWord[LEN];
	auto int wolen;
	auto int ctr;
	auto int Pscore = 0;

	grid.Print(Pscore);
	//loop for words + print
	while(cin >> userWord)
		{
		wolen = strlen(userWord);
		for(ctr=0; ctr<wolen; ctr++) userWord[ctr] = toupper(userWord[ctr]);
		//if !, print all words
		if('!' == userWord[0])
			{
			cout << "  (M)y Words\n  (C)ustom\n  (S)olve\n  (N)ew\n  (Q)uit\n";
			cin >> datChar;
			switch(toupper(datChar))
				{
//create case to see pList
				case 'M':
					cout << "\n Your words:\n";
					tbag = pList;
					while(tbag)
						{
						cout << "  " << tbag->word << endl;
						tbag = tbag->next;
						}
					break;
				case 'C':
					grid.Customize();
					DestroyList(solvedList);
					DestroyList(pList);
					pList = NULL;
					solvedList = grid.FindWords(wordList);
					Pscore = 0;
					break;
				case 'S':
					tbag = solvedList;
					while(tbag)
						{
						cout << tbag->word << endl;
						tbag = tbag->next;
						}
					cout << "Max possible score: " << grid.maxP << endl;
					break;
				case 'N':
					grid.Randomize(dice);
					DestroyList(solvedList);
					DestroyList(pList);pList = NULL;
					solvedList = grid.FindWords(wordList);
					Pscore = 0;
					break;
				case 'Q':
					cout << "Thank you for playing!\n\n";
					exit(EXIT_SUCCESS);
				}
			}
		else{
			if (checkGuess(userWord, solvedList, pList))
				{
				pList = addWord(userWord, pList);
				cout << endl << userWord << " found!\n";
				}
			else cout << "Sorry, invalid guess\n";
			}

		Pscore = MaxScore(pList);
		grid.Print(Pscore);
		}//should keep looping until the menu calls exit

	DestroyList(wordList);
//release lists... bah
	return 0;
}


int Boogle::Randomize(const char* dice[])
{
	auto int sand;
	auto int ctr;

	for (ctr = 0; ctr < 16; ctr++)
		{
		sand = rand() % 6;
		letters[ctr] = dice[ctr][sand];
		}
	letters[ctr] = '\0';
	return ctr;
}


int Boogle::Customize(void)
{
	auto char inch;
	auto int ctr;

	cout << "Enter 16 letters:\n";
	for (ctr = 0; ctr < 16; ctr++)
		{
		cin >> inch;
		letters[ctr] = toupper(inch);
		}
	letters[ctr] = '\0';
	return ctr;
}



int Boogle::Print(int score)
{
	auto int ctr;

	for (ctr = 0; ctr < 16; ctr++)
		{
		if (0 == ctr%4) cout << "\n  ";
		if ('Q' == letters[ctr]) cout << " Qu";
		else cout << " " << letters[ctr] << " ";
		}
	cout << "\nScore: " << score << "\nEnter a word: (Enter an '!' for menu)\n";
	return ctr;
}


bool isIn(int a, const int* path, int n)
{
	auto int ctr;
	for (ctr = 0; ctr < n; ctr++)
		{
		if (a == path[ctr]) return true;
		}
	return false;
}

int nextAdj(const int* path, int n)
{
	auto int ctr = 0;
	auto int numPoss;
	auto int poss[8];

	switch(path[n-1])
		{
		case 0: poss[0] = 1;
				poss[1] = 4;
				poss[2] = 5;
				numPoss = 3;
				break;
		case 1: poss[0] = 0;
				poss[1] = 2;
				poss[2] = 4;
				poss[3] = 5;
				poss[4] = 6;
				numPoss = 5;
				break;
		case 2: poss[0] = 1;
				poss[1] = 3;
				poss[2] = 5;
				poss[3] = 6;
				poss[4] = 7;
				numPoss = 5;
				break;
		case 3: poss[0] = 2;
				poss[1] = 6;
				poss[2] = 7;
				numPoss = 3;
				break;
		case 4: poss[0] = 0;
				poss[1] = 1;
				poss[2] = 5;
				poss[3] = 8;
				poss[4] = 9;
				numPoss = 5;
				break;
		case 5: poss[0] = 0;
				poss[1] = 1;
				poss[2] = 2;
				poss[3] = 4;
				poss[4] = 6;
				poss[5] = 8;
				poss[6] = 9;
				poss[7] = 10;
				numPoss = 8;
				break;
		case 6: poss[0] = 1;
				poss[1] = 2;
				poss[2] = 3;
				poss[3] = 5;
				poss[4] = 7;
				poss[5] = 9;
				poss[6] = 10;
				poss[7] = 11;
				numPoss = 8;
				break;
		case 7: poss[0] = 2;
				poss[1] = 3;
				poss[2] = 6;
				poss[3] = 10;
				poss[4] = 11;
				numPoss = 5;
				break;
		case 8: poss[0] = 4;
				poss[1] = 5;
				poss[2] = 9;
				poss[3] = 12;
				poss[4] = 13;
				numPoss = 5;
				break;
		case 9: poss[0] = 4;
				poss[1] = 5;
				poss[2] = 6;
				poss[3] = 8;
				poss[4] = 10;
				poss[5] = 12;
				poss[6] = 13;
				poss[7] = 14;
				numPoss = 8;
				break;
		case 10:poss[0] = 5;
				poss[1] = 6;
				poss[2] = 7;
				poss[3] = 9;
				poss[4] = 11;
				poss[5] = 13;
				poss[6] = 14;
				poss[7] = 15;
				numPoss = 8;
				break;
		case 11:poss[0] = 6;
				poss[1] = 7;
				poss[2] = 10;
				poss[3] = 14;
				poss[4] = 15;
				numPoss = 5;
				break;
		case 12:poss[0] = 8;
				poss[1] = 9;
				poss[2] = 13;
				numPoss = 3;
				break;
		case 13:poss[0] = 8;
				poss[1] = 9;
				poss[2] = 10;
				poss[3] = 12;
				poss[4] = 14;
				numPoss = 5;
				break;
		case 14:poss[0] = 9;
				poss[1] = 10;
				poss[2] = 11;
				poss[3] = 13;
				poss[4] = 15;
			 	numPoss = 5;
			 	break;
		case 15:poss[0] = 10;
				poss[1] = 11;
				poss[2] = 14;
				numPoss = 3;
				break;
		default: cout << "path[n-1] out of bounds in nextAdj. n = " << n << endl;
				int j; for (j=0; path[j] > 0; j++) cout << path[j] << " ";
		}

	//basically the NEXT possible...
	while (ctr<numPoss && ((path[n] >= poss[ctr]) || (isIn(poss[ctr], path, n-1)))) ctr++;

	//poss[ctr] is our nextAdj, unless there was no poss.
	if (ctr == numPoss) return -1;
	else return poss[ctr];
}


//major work in progress here
WNode* Boogle::FindWords(WNode* dict)
{
	auto WNode* Words = NULL;
	auto char bogWord[17];
	auto int ltrCpy;
	auto int path[16] = {0, 1, 2, -1};
	auto int n = 3;
	auto int wlen;
	auto int val;
	auto bool bLoop = true;


	while(bLoop)//this guy rotates from the end
		{
		for(ltrCpy = 0; ltrCpy < n; ltrCpy++)
			{
			bogWord[ltrCpy] = letters[(path[ltrCpy])];
			}
		bogWord[ltrCpy] = '\0';

		val = validate(bogWord, dict, Words);

		if		(val == 1) Words = addWord(bogWord, Words);
		else if (val <  0) n = -1 * val;


//path[n] should be set to the next adjecent available.
//if there is no possible next path, roll back the path and...
//look for a next possible. again if there is none, roll back
//if we roll back to n=0, we should increment path[0] until it is >15
		while (-1 == (path[n] = nextAdj(path, n)))
			{
			n--;
			if (0 == n)
				{
				path[0]++;
				if (path[0] == 16)
					{
					bLoop = false;
					break;
					}
				n++;
				}
			}
		++n;
		path[n] = -1;
		}

	maxP = MaxScore(Words);
	return Words;
}




WNode* Boogle::FakeWords(WNode* dict)
{
	auto WNode* Words = NULL;
	auto char word[17];
	auto int ctr,ctr2, val;
	for (ctr=0; ctr<3; ctr++)
		{
		cout << "Enter a fake word: ";
		cin >> word;
		val = validate(word, dict, Words);
		if (-1 == val) cout << "Invalid\n";
		else if (1 == val) Words = addWord(word, Words);
		}
	maxP = MaxScore(Words);
	return Words;
}


int validate(const char* word, WNode* dict, WNode* prev)
{//-n invalid word, 0 possible word, 1 add to list

	auto int wlen = strlen(word);

	while(prev)
		{//may have been gettting errors here...
		if (wlen == strlen(prev->word))//may still be...
			{
			if (0 == strcmp(word, prev->word)) return wlen * -1;
			}
		prev = prev->next;
		}

	auto WNode* traverser = dict;
	auto int ctr = 0;

	while (traverser)
		{
		while (traverser->word[ctr] < word[ctr])
			{
			if (ctr > 0) if (traverser->word[ctr-1] != word[ctr-1]) break;
			traverser = traverser->next;
			}
		if (traverser->word[ctr] == word[ctr]) ctr++;
		else return -1 * ctr;

		if (ctr == wlen)
			{
			if (strlen(traverser->word) == wlen)
				{
				return 1;
				}
			else return 0;
			}
		}

}



WNode* addWord(const char* word, WNode* list)
{
	auto int wlen = strlen(word)+1;
	auto WNode* adder = new WNode;

	strncpy(adder->word, word, wlen);
	adder->next = list;

	return adder;
}

File: boggle.py
#Simple Boggle script

import random
from random import shuffle
random.seed()



def makeTrie(*Wordlist):
	root = dict()
	for Words in Wordlist:
		for word in Words:
			currentDict = root
			for letter in list(word):
				currentDict = currentDict.setdefault(letter, {})
			currentDict = currentDict.setdefault('_x', '_x')
	return root
	

def inTrie(trie, word):
	currentDict = trie
	for letter in list(word):
		if letter in currentDict:
			currentDict = currentDict[letter]
		else:
			return False
	else:
		if '_x' in currentDict:
			return True
		else:
			return False




def NumCharMatch(trie, word):
	currentDict = trie
	match = 0
	for letter in word:
		if letter in currentDict:
			match += 1
			currentDict = currentDict[letter]
	return match



def nextAdj(path, n):
	poss = []
	if path[n-1] == 0:
		poss = [1,4,5]
	elif path[n-1] == 1:
		poss = [0,2,4,5,6]
	elif path[n-1] == 2:
		poss = [1,3,5,6,7]
	elif path[n-1] == 3:
		poss = [2,6,7]
	elif path[n-1] == 4:
		poss = [0,1,5,8,9]
	elif path[n-1] == 5:
		poss = [0,1,2,4,6,8,9,10]
	elif path[n-1] == 6:
		poss = [1,2,3,5,7,9,10,11]
	elif path[n-1] == 7:
		poss = [2,3,6,10,11]
	elif path[n-1] == 8:
		poss = [4,5,9,12,13]
	elif path[n-1] == 9:
		poss = [4,5,6,8,10,12,13,14]
	elif path[n-1] == 10:
		poss = [5,6,7,8,11,13,14,15]
	elif path[n-1] == 11:
		poss = [6,7,10,14,15]
	elif path[n-1] == 12:
		poss = [8,9,13]
	elif path[n-1] == 13:
		poss = [8,9,10,12,14]
	elif path[n-1] == 14:
		poss = [9,10,11,13,15]
	elif path[n-1] == 15:
		poss = [10,11,14]

#	print n, path	
	if path[n] == poss[-1]:
		return -1
	else:
		i = 0
		while (path[n] >= poss[i] or poss[i] in path[:n]):
			i += 1
			if i == len(poss):
				return -1
		return poss[i]



class Boggle:
	letters = []
	
	def __init__(self, dice):
		self.letters = []
		dies = []
		for i in xrange(16):
			dies.append(i)
		shuffle(dies)
		for i in dies:
			self.letters.append(dice[i][random.randrange(0, 6)])

	def Pgrid(self, score):
		print " ",
		for i in xrange(16):
			if i in (4, 8, 12):
				print "\n ",
			if 'Q' == self.letters[i]:
				print 'Qu',
			else:
				print self.letters[i] + ' ',
			if 15 == i:
				print "\nCurrent score is ", score



	def oldPgrid(self, score):
		print ""
		print self.letters[:4]
		print self.letters[4:8]
		print self.letters[8:12]
		print self.letters[12:]
		print "Current score is ", score
		print ""
	
	def Solve(self, Dtrie):
		path = [0, 1, 2, -1,0,0,0,0,0,0,0,0,0,0,0,0]
		n = 3
		looping = True
		bog = self.letters[:]
		pathWord = []
		solution = []
		pWord = ""
		
		while(looping):
			pathWord = []
			i = 0
			while i < n:
				if 'Q' == bog[path[i]]:
					pathWord.append('QU')
				else:
					pathWord.append(bog[path[i]])
				i += 1
			
			pWord = "".join(pathWord)
			if inTrie(Dtrie, pWord) and pWord not in solution:
				solution.append(pWord)
#				print "Adding ", pWord
			else:
				if 'Q' in list(pWord):
					n = NumCharMatch(Dtrie, pWord) - 1
				else:
					n = NumCharMatch(Dtrie, pWord)

			#set next adjecent to path[n]
			Abool = True
			while (Abool):
				path[n] = nextAdj(path, n)
				if (-1 != path[n]):
					Abool = False
					break
				n -= 1
				if 0 == n:
					path[0] += 1
					if 16 == path[0]:
						looping = False
						break
					n += 1
					path[n] = -1
			n += 1
			path[n] = -1
		
		
		return solution
			
			

	def Custom(self):
		lString = raw_input("Enter 16 letters: ")
		self.letters = list(lString.upper())
		
	




def main():
	dice = ("AAEEGN", "ELRTTY", "AOOTTW", "ABBJOO",
	"EHRTVW", "CIMOTU", "DISTTY", "EIOSST",
	"DELRVY", "ACHOPS", "HIMNQU", "EEINSU",
	"EEGHNW", "AFFKPS", "HLNNRZ", "DEILRX")
	
	with open('dict.txt') as file:
	    Words = file.read().splitlines()
	
	Dtrie = makeTrie(Words)

	grid = Boggle(dice);
	solvedWords = grid.Solve(Dtrie)
	
	scoreCtr = 0
	grid.Pgrid(scoreCtr)
	playerWords = []
	
	#print grid, loop and ask for words
	while (True):
		guess = raw_input("Enter a word ('!' for menu): ")
		if '!' == guess:
			print " Menu:"
			print "  (M)y words"
			print "  (C)ustom board"
			print "  (S)olve"
			print "  (N)ew game"
			print "  (Q)uit"
			choice = raw_input("Please make a selection: ")
			choice = choice.upper()
			
			if 'Q' == choice:
				print "Thank you for playing!"
				exit()
			elif 'M' == choice:
				print ""
				print "Your words: ", playerWords
			elif 'S' == choice:
				print solvedWords
				maxScore = 0
				for word in solvedWords:
					maxScore += len(word) - 2
				print "Max score = ", maxScore
			elif 'N' == choice:
				grid.__init__(dice)
				solvedWords = grid.Solve(Dtrie)
				scoreCtr = 0
				playerWords = []
			elif 'C' == choice:
				grid.Custom()
				solvedWords = grid.Solve(Dtrie)
				scoreCtr = 0
				playerWords = []

		else:
			guess = guess.upper()
			if guess in solvedWords and guess not in playerWords:
				playerWords.append(guess)
				print guess, "found!"
				scoreCtr += len(guess) - 2
			else:
				print "Sorry, invalid guess"
		
		grid.Pgrid(scoreCtr)
		
		
	#delete lists and allocated memory
		


main()
